File: wot-dijkstra/src/main/java/org/pgpainless/wot/dijkstra/Cost.java
package org.pgpainless.wot.dijkstra;

public class Cost {

    public static class SimpleCost extends Cost {

        private final double weight;

        public SimpleCost(double weight) {
            this.weight = weight;
        }

        public double getWeight() {
            return weight;
        }

        @Override
        public String toString() {
            return Double.toString(getWeight());
        }

        @Override
        public boolean equals(Object obj) {
            if (obj == null) {
                return false;
            }
            if (this == obj) {
                return true;
            }
            if (!(obj instanceof SimpleCost)) {
                return false;
            }
            SimpleCost other = (SimpleCost) obj;
            return getWeight() == other.getWeight();
        }

        @Override
        public int hashCode() {
            return (int) getWeight();
        }
    }

    public static class TrustCost extends Cost {

        private final int depth;
        private final int amount;
        private final String regex;

        public TrustCost(int depth, int amount, String regex) {
            this.depth = depth;
            this.amount = amount;
            this.regex = regex;
        }

        public int getDepth() {
            return depth;
        }

        public int getAmount() {
            return amount;
        }

        public String getRegex() {
            return regex;
        }

        @Override
        public String toString() {
            return "d=" + getDepth() + ",a=" + getAmount() + (regex == null ? "" : ",r=" + getRegex());
        }
    }
}

File: wot-dijkstra/src/main/kotlin/org/pgpainless/wot/query/Cost.kt
// SPDX-FileCopyrightText: 2023 Heiko Schäfer <heiko@schaefer.name>, 2022-2023, pep foundation
//
// SPDX-License-Identifier: LGPL-2.0-only

package org.pgpainless.wot.query

import kotlin.math.min

/**
 * A path's "cost".
 *
 * This Cost type is used for comparison in Dijkstra's algorithm.
 */
internal class Cost(
        // The path's length (i.e., the number of hops to the target).
        // *Less* length is "cheaper" (short paths require less "depth" in delegations).
        val length: Int,

        // The trust amount along this path.
        // Smaller trust amount is "more expensive"
        // (paths with a low trust amount are more constraining and thus less desirable).
        val amount: Int,
) : Comparable<Cost> {

    // "Smaller than" means: the path is "cheaper", and thus preferable:
    // - A small length (requiring fewer hops).
    // - For equal length: A higher "trust amount" (which is less constraining)
    override fun compareTo(other: Cost) =
            compareValuesBy(this, other, { it.length }, { -it.amount })

    override fun toString() = "length $length, amount $amount"

    /** Calculate the cost of a path,  when adding a new segment with trust amount `amount` */
    fun extendBy(amount: Int) = Cost(this.length + 1, min(amount, this.amount))

}